src/Text/Pandoc/Readers/Vimwiki.hs

```diff
@@ -0,0 +1,398 @@
+    * [X] tag
+    * [ ] sub- and super-scripts
+* misc:
+    * [X] `TODO:` mark
+    * [ ] placeholders: %title and %date -> metadata, %template -> template
+--}
+
+module Text.Pandoc.Readers.Vimwiki ( readVimwiki
+                                 ) where
+import Control.Monad.Except (throwError)
+import Control.Monad (guard)
+import Data.Default 
+import Data.Functor.Identity
+import Data.Maybe
+import Data.List (isInfixOf)
+import Data.List.Split (splitOn)
+import Data.Text (strip)
+import Text.Pandoc.Builder (Blocks, Inlines, trimInlines, fromList, toList)
+import qualified Text.Pandoc.Builder as B (doc, toList, headerWith, str, space, strong, emph, strikeout, code, link, image, spanWith, math, para, horizontalRule, blockQuote, codeBlock, displayMath, bulletList, plain, orderedList, simpleTable, softbreak)
+import Text.Pandoc.Class (PandocMonad, report, PandocIO, runIO)
+import Text.Pandoc.Definition (Pandoc, nullAttr, Inline(Space), Block(BulletList, OrderedList))
+import Text.Pandoc.Error (PandocError)
+import Text.Pandoc.Logging (LogMessage(ParsingTrace))
+import Text.Pandoc.Options (ReaderOptions)
+import Text.Pandoc.Parsing (readWithM, ParserT, stateOptions, ParserState, blanklines, registerHeader, spaceChar, stateAllowLinks, emailAddress, guardEnabled, uri)
+import Text.Parsec.Char (spaces, char, anyChar, newline, string, noneOf)
+import Text.Parsec.Error (ParseError)
+import Text.Parsec.Combinator (eof, choice, many1, manyTill, count, skipMany1, notFollowedBy)
+import Text.Parsec.Pos (sourceColumn)
+import Text.Parsec.Prim (many, getPosition, try, runParserT)
+import Text.Parsec.String (Parser)
+import Text.Parsec (parse)
+import Text.Parsec.Char (oneOf, space)
+import Text.Parsec.Combinator (lookAhead, between)
+import Text.Parsec.Prim ((<|>), (<?>), skipMany)
+import Text.Pandoc.Options (Extension(Ext_autolink_bare_uris))
+
+readVimwiki :: PandocMonad m => ReaderOptions -> String -> m Pandoc
+readVimwiki opts s = do
+  res <- readWithM parseVimwiki def{ stateOptions = opts } s
+  case res of
+       Left e -> throwError e
+       Right result -> return result
+
+type VwParser = ParserT [Char] ParserState
+
+
+-- constants
+
+specialChars :: [Char]
+specialChars = "=*-#[]_~{`$|:%"
+
+spaceChars :: [Char] 
+spaceChars = " \t\n"
+
+-- main parser
+
+parseVimwiki :: PandocMonad m => VwParser m Pandoc
+parseVimwiki = do
+  bs <- mconcat <$> many block
+  spaces
+  eof
+  return $ B.doc bs
+
+-- block parser
+
+block :: PandocMonad m => VwParser m Blocks
+block = do
+  pos <- getPosition
+  res <- choice [ mempty <$ blanklines
+                , header
+                , hrule
+                , mempty <$ comment
+                , mixedList
+                , blockQuote
+                , preformatted
+                , displayMath 
+                , table
+                , para
+                ]
+  report $ ParsingTrace (take 60 $ show $ B.toList res) pos
+  return res
+
+header :: PandocMonad m => VwParser m Blocks
```

Comment:
Stylistically, we prefer to have the type signatures right above the function definitions.